src/async.ts

```diff
@@ -19,7 +19,7 @@ export const delay = (ms: number): Fx<Delay & Async, void> => doFx(function* ()
   return yield* delay(ms)
 })
 
-export const timeout = <C extends Async, A>(ms: number, c: Fx<C, A>): Fx<C & Delay & Fail, A> =>
+export const timeout = <C, A>(ms: number, c: Fx<C, A>): Fx<C & Async & Delay & Fail, A> =>
```

Comment:
It would be nice to find a way to reinstate these effect constraints.  As it stands, they end up limiting effect polymorphism in applications, in ways that aren't intuitive.

Reply:
How do they limit effect polymorphism?

Reply:
It's definitely hard to see, but I'll try to walk through it.

1. If timeout constrains C to `extends Async`, then [calling it here](https://github.com/briancavalier/fx-ts/pull/85/files/6245903cc11cd1f837e549428bc2a6f2fb0ee10b#diff-3791aad6a20cf56f11de855f9c647d39R25) requires getLocation's effects to have the same constraint.
2. That, in turn, constrains `Effects` to be `extends Async`, and the constraint propagates all the way out to the top level.
3. At that point, the application may only be run with a GetLocation that is Async, even if it were possible to provide a GetLocation that isn't Async, for example: `getLocation = (ip: string) => pure(defaultLocation)` would be rejected by the compiler.

It could be that I'm expecting too much freedom, and assuming GetLocation must at least be Async is fine.  Similarly, though, it'd have to be `extends Fail`, since it's reasonable to have the [`extends Fail` constraint on `catchAll`](https://github.com/briancavalier/fx-ts/pull/85/files/6245903cc11cd1f837e549428bc2a6f2fb0ee10b#diff-a72f5aadd00b4b390e5e6add160e22c1R15).

Reply:
I think the constraints are perfectly fine, but, yes, it would be great to have some mechanism to  reinstate the constraints. To me, it feels like the same constraints an `IO` type imposes.

Reply:
Would it be possible to wrap it in an `IO`  type, e.g., `const timeout = (ms, c) => () => race(c, delayFail(ms))`, andd avoid the constraint?

Reply:
Hmm, I don't think so, since the return type of that IO thunk is still `Fx<Effects, A>`.  These two goals definitely seem at odds right now:

1. An unconstrained application Effects parameter
2. Functions with constraints on the effects

It might be reasonable just to provide a type alias for `Async & Fail`, since it will tend to be fairly common.  I think [koka does something similar](https://github.com/koka-lang/koka/blob/ad375f8030569ab4289c94eab45c2d8ba3aee2be/lib/std/async.kk#L45).

```ts
export type AsyncFail = Async & Fail
```

In that case the domain might define:

```ts
export type GetPets<Effects extends AsyncFail> = (l: GeoLocation, radiusMiles: number) => Fx<Effects, Pets>

export type GetLocation<Effects extends AsyncFail> = (host: string) => Fx<Effects, Location>
```